### PR TITLE
[API-1362] Do not interleave async_write calls 

### LINKED ISSUE #980: Hazelcast C++ API issue with large messages
Reported by community:
[hazelcast_example.txt](https://github.com/hazelcast/hazelcast-cpp-client/files/8861984/hazelcast_example.txt)
[hazelcast_frame_buffer_log.txt](https://github.com/hazelcast/hazelcast-cpp-client/files/8861987/hazelcast_frame_buffer_log.txt)

C++ compiler version: 
Hazelcast Cpp client version: 4.2
Hazelcast server version: 4.2
Number of the clients: 
Cluster size, i.e. the number of Hazelcast cluster members:
OS version (Windows/Linux/OSX):

Please attach relevant logs and files for client and server side.

#### Actual behaviour
Encountered an issue with publishing of large messages from Hazelcast C++ API.  Our calculation nodes publish messages 100KB to 200KB in size using the C++ API.  Occasionally, these messages arrive at the Hazelcast cluster in a garbled form.  It turns out that they are interleaved with data from heartbeat / ping messages.

I believe it is due to a heartbeat messages and data being written from separate threads.  This may not be a problem for smaller messages (<65K) because each write is atomic. For larger messages, boost::asio library splits them into smaller chunks resulting in interleaving with ping messages.

ttaching an example program that we ran on our systems that reproduced the problem. This generates a large payload and repeatedly publishes it to different keys in a map. We are setting it to time out if the publish takes longer than 30 seconds – and we got timeouts after about 70 publishes. In the test code, we also increased the heartbeat to happen every 10ms in order to force the problem to happen more frequently. We are using hazelcast-cpp-client 4.2.0, and hazelcast node 4.2.

 
Attaching a log snippet from the hazelcast node during one instance of this problem. As the data is passed into hazelcast, everything up through frame 3 is received correctly. However, once frame 4 arrives, the hazelcast node interprets the frame buffer length to be an enormously large number. It then incorrectly parses the future data received as if it were all part of frame 4. In the course of our investigation, we discovered that this happens because a ping message is interleaved into frame 3, making it longer than expected. We were able to identify the location of the ping message in one case, remove it from the stream of bytes received, and correctly deserialize the data that was expected to be in frame 3.

#### Steps to reproduce the behaviour
Please run the attached C++ code for reproducing.

### PR DESCRIPTION
**Pull request by ihsandemir**
Fixes the async io write logic. We need to have only one outstanding async_write operation for a single socket. Failing to do this causes interleaving the write requests and causing corrupt messages at the server side. See [this discussion](https://itecnote.com/tecnote/c-boost-asio-async_write-how-to-not-interleaving-async_write-calls/) The write requests needs to be queued and consumed either on the async_write callback handler or directly on write requests if no outstanding write is in progress.

fixes #980

### CODE REVIEW

**Comment by yuce** [comment]
@ihsandemir Tests for Linux and Windows seem to be failing.

**Comment by ihsandemir** [comment]
> @ihsandemir Tests for Linux and Windows seem to be failing.

[This](https://github.com/hazelcast/hazelcast-cpp-client/pull/986) other PR should fix the failure problem.

File: hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h [hunk 1]
```diff
@@ -90,54 +90,45 @@ namespace hazelcast {
...
+                                correlation_id composed_id;
+                            } c_id_union;
+                            c_id_union.composed_id = {connection->get_connection_id(), call_id};
+                            message_call_id = c_id_union.id;
+                            success = connection->invocations.insert({message_call_id, invocation}).second;
+                            if (success) {
+                                message->set_correlation_id(c_id_union.id);
+                            }
+
+                            //message->print_frame_sizes();
```
Comment (srknzl) [inline hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h:116]: If this is a leftover we can delete this

File: hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h [hunk 2]
```diff
@@ -90,54 +90,45 @@ namespace hazelcast {
                                     const std::shared_ptr<spi::impl::ClientInvocation> invocation) override {
                         check_connection(connection, invocation);
                         auto message = invocation->get_client_message();
-                        socket_strand_.post([=]() {
+                        socket_strand_.post([connection, invocation, message, this]() {
```
Comment (srknzl) [inline hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h:93]: Why this change is required? What is difference between [=]  and [variables]?

File: hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h [hunk 3]
```diff
@@ -226,10 +217,36 @@ namespace hazelcast {
...
+                                auto message = (boost::format{
+                                        "Error %1% during invocation write for %2% on connection %3%"} %
+                                                ec % *invocation % *connection).str();
+                                connection->close(message);
+                            } else {
+                                // update the connection write time
+                                connection->last_write_time(std::chrono::steady_clock::now());
+
+                                if (!this->outbox_.empty()) {
+                                    do_write(connection, invocation);
```
Comment (srknzl) [inline hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h:236]: This handler can't run in parallel with `socket.strand.post()` handler above, right? Are they executed in the same thread?

**Review by srknzl — COMMENTED** [review]
Looks good in general though I am not familiar with these at all.

**Comment by ihsandemir** [comment]
> @ihsandemir Tests for Linux and Windows seem to be failing.

I sent the [other PR](https://github.com/hazelcast/hazelcast-cpp-client/pull/986) which will fix build failures. This was an old branch which needed updates to docker files from master.

File: hazelcast/include/hazelcast/client/protocol/ClientMessage.h [hunk 4]
```diff
@@ -1020,8 +1020,8 @@ namespace hazelcast {
                 std::string operation_name_;
 
                 std::vector<std::vector<byte>> data_buffer_;
-                size_t buffer_index_;
-                size_t offset_;
+                size_t buffer_index_ {0};
+                size_t offset_ {0};
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/protocol/ClientMessage.h:1024]: They are incremented after this during client message write methods (see method `peek`). C++ does not zero initialize by default (probably due to performance), and it can be any arbitrary value actually depending on the memory region.

File: hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h (hunk 1 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h:116]: yes, leftover. Removed.

File: hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h (hunk 2 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h:93]: This makes the variables used more explicit. Otherwise, the other version is implicit and you can easily be coping variables you did not intend copy in the first place.  We switched to this usage after getting more experience with lambdas.

File: hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h (hunk 3 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h:236]: We execute handler using this wrapper `socket_strand_.wrap(handler)` which guarantees that the handler is executed using the same strand. Strand guarantees that the codes using the same strands can not run in parallel, they will be sequential through the strand. see asio strand concept. 

For our implementation, yes they are at the same io thread. As a matter of fact we only have a single io thread.

File: hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h [hunk 5]
```diff
@@ -90,54 +90,45 @@ namespace hazelcast {
...
                         auto message = invocation->get_client_message();
-                        socket_strand_.post([=]() {
+                        socket_strand_.post([connection, invocation, message, this]() {
                             if (!check_connection(connection, invocation)) {
                                 return;
                             }
 
                             bool success;
                             int64_t message_call_id;
-                            do {
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h:100]: I put this back. That loop is still needed for the very little possibility that the generated call ids clash. This can only happen usually if it is an event they stay in the map for a long time and if the counter overflows back to the original number.

File: hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h [hunk 6]
```diff
@@ -90,54 +90,45 @@ namespace hazelcast {
...
+                            struct correlation_id {
+                                int32_t connnection_id;
+                                int32_t call_id;
                             };
+                            union {
+                                int64_t id;
+                                correlation_id composed_id;
+                            } c_id_union;
+                            c_id_union.composed_id = {connection->get_connection_id(), call_id};
+                            message_call_id = c_id_union.id;
```
[the remark this answers is not in the document]
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h:110]: It can be done. Endianness may not be a problem since the correlation id is only unique per connection and we always use the incremented `call_id_counter_`. Does that look simpler? Is that one more efficient or current impl? I am not sure, let me try to evaluate. I can make it a method call, yes.

[the remark this answers is not in the document]
Comment (yuce) [reply hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h]: I've checked both options on the compiler explorer: https://godbolt.org/z/5ffdjcob8

I am definitely not an expert in C++ or assembly, but I believe the shift + OR could be very slightly more efficient (my reasoning is, there are less stack MOV instructions with the former) but I don't think that's important at all.

[the remark this answers is not in the document]
Comment (yuce) [reply hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h]: Just a note using `-O1` with GCC optimizes both approaches to the same code: https://godbolt.org/z/ncYaeMMqE

[the remark this answers is not in the document]
Comment (yuce) [reply hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h]: Surprisingly MSVC optimizes the struct+union one a lot better: https://godbolt.org/z/nT3Kre9fh

File: hazelcast/include/hazelcast/client/protocol/ClientMessage.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply hazelcast/include/hazelcast/client/protocol/ClientMessage.h:1024]: But is this related to this PR? So it was a bug I guess.

**Review by srknzl — APPROVED** [review]
